File: compiler/nodes/expressions/do.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type DoExpressionNode struct {
	Input database.Node
	Facts *database.Facts
}

func (node *DoExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseDoExpression(parser *syntax.Parser) (*DoExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("DoKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this `do` expression")

	value, err := ParseAtomicExpression(parser)
	if err != nil {
		return nil, err
	}

	return &DoExpressionNode{
		Input: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *DoExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	visitor.Visit(node.Input)
	visitor.Constraint(typecheck.NewTypeConstraint(node.Input, typecheck.BlockType(node)))
}

func (node *DoExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)

	c.WriteString(span, "await (")
	if err := c.Write(node.Input); err != nil {
		return err
	}
	c.WriteString(span, ")()")

	return nil
}

File: compiler/src/nodes/expressions/do.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{parse_atomic_expression, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct DoExpressionNode {
    pub input: NodeRef,
}

impl Node for DoExpressionNode {}

pub fn parse_do_expression(parser: &mut Parser<'_>) -> Result<DoExpressionNode, ParseError> {
    parser.token(TokenKind::DoKeyword)?;
    parser.commit("in this `do` expression");

    let input = parse_atomic_expression(parser)?;

    Ok(DoExpressionNode { input })
}

impl Visit for DoExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        visitor.visit(&self.input);
        visitor.edge(&self.input, node, "input");

        visitor.constraint(TypeConstraint::new(
            self.input.clone(),
            visitor.block_type(node.clone()),
        ));
    }
}

impl Codegen for DoExpressionNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        ctx.write_string("await (");
        ctx.write(&self.input)?;
        ctx.write_string(")()");

        Ok(())
    }
}
